daemon/logger/jsonfilelog/read.go

```diff
@@ -121,96 +119,106 @@ func tailFile(f io.ReadSeeker, logWatcher *logger.LogWatcher, tail int, since ti
 	}
 }
 
-func followLogs(f *os.File, logWatcher *logger.LogWatcher, notifyRotate chan interface{}, since time.Time) {
-	dec := json.NewDecoder(f)
-	l := &jsonlog.JSONLog{}
+func (l *JSONFileLogger) followLogs(f *os.File, logWatcher *logger.LogWatcher, notifyRotate chan interface{}, since time.Time) {
+	var (
+		msg        *logger.Message
+		err        error
+		rotated    bool
+		readIndex  uint64
+		writeIndex uint64
+
+		dec        = json.NewDecoder(f)
+		log        = &jsonlog.JSONLog{}
+		subscriber = l.writeNotifier.Subscribe()
+		watchClose = logWatcher.WatchClose()
+	)
+
+	go func() {
+		for {
+			if _, ok := <-subscriber; !ok {
```

Comment:
But you don't need to read per write, you can get a write notification and read until EOF.

Alternatively you can use a much larger buffer on the channel (e.g., 1000) and continue reading per write.